### PR TITLE
Fix some kinds of invalid polygons

### PR DESCRIPTION
**Pull request by zerebubuth**
Good news: It passes the tests, including many more types of deliberately invalid polygons. Also includes several failures noticed in the `dev` cluster logs.

Bad news: It can't quite do everything, and it's pretty slow. I hope that it doesn't matter too much, as the code is only run on polygons which are invalid. Also, at the moment it drops:

* Inner rings which are smaller than 0.5 square units. Given that the smallest valid polygon in integer coordinates has area 0.5, we can probably live without these.
* Inner rings which would make the polygon invalid. This is [possibly a bug](https://trac.osgeo.org/geos/ticket/789), but does happen sometimes. The symptoms might be missing islands or lakes, missing courtyards in buildings. We'll have to investigate these as we find them to see if we can work around it.

@iandees & @rmarianski could you review, please?

### CODE REVIEW

File: mapbox_vector_tile/polygon.py [hunk 1]
```diff
@@ -43,34 +45,134 @@ def _coords(shape):
...
+            # that's valid than nothing.
+            if diff.is_valid:
+                poly = diff
+
+        assert poly.is_valid
+        polygons.append(poly)
+        children = []
+
+    else:
+        assert len(children) == 0
```
Comment (iandees) [inline mapbox_vector_tile/polygon.py]: It might be good to add a tiny comment about why you're making this assertion.

**Comment by nvkelso** [comment]
Tiles sync: let's get this reviewed ASAP and get a new dev build.

File: mapbox_vector_tile/polygon.py (hunk 1 above)
Comment (zerebubuth) [inline mapbox_vector_tile/polygon.py]: Good idea! I've added something in 231011b which hopefully makes it clearer why that's there.

Comment (rmarianski) [inline mapbox_vector_tile/polygon.py]: :+1:

**Pull request merged**